btcec/schnorr/musig2/nonces.go

```diff
@@ -0,0 +1,209 @@
+		// By default, we always use the crypto/rand reader, but the
+		// caller is able to specify their own generation, which can be
+		// useful for deterministic tests.
+		randReader: rand.Read,
+	}
+}
+
+// GenNonces generates the secret nonces, as well as the public nonces which
+// correspond to an EC point generated using the secret nonce as a private key.
+func GenNonces(options ...NonceGenOption) (*Nonces, error) {
+	opts := defaultNonceGenOpts()
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	// Generate two 32-byte random values that'll be the private keys to
+	// the public nonces.
+	var k1, k2 [32]byte
+	if _, err := opts.randReader(k1[:]); err != nil {
+		return nil, err
+	}
+	if _, err := opts.randReader(k2[:]); err != nil {
+		return nil, err
+	}
+
+	var nonces Nonces
+
+	var k1Mod, k2Mod btcec.ModNScalar
+	k1Mod.SetBytes(&k1)
+	k2Mod.SetBytes(&k2)
+
+	// The secret nonces are serialized as the concatenation of the two 32
+	// byte secret nonce values.
+	k1Mod.PutBytesUnchecked(nonces.SecNonce[:])
+	k2Mod.PutBytesUnchecked(nonces.SecNonce[btcec.PrivKeyBytesLen:])
+
+	// Next, we'll generate R_1 = k_1*G and R_2 = k_2*G. Along the way we
+	// need to map our nonce values into mod n scalars so we can work with
+	// the btcec API.
+	nonces.PubNonce = secNonceToPubNonce(nonces.SecNonce)
+
+	return &nonces, nil
+}
+
+// AggregateNonces aggregates the set of a pair of public nonces for each party
+// into a single aggregated nonces to be used for multi-signing.
+func AggregateNonces(pubNonces [][PubNonceSize]byte) ([PubNonceSize]byte, error) {
+	// combineNonces is a helper function that aggregates (adds) up a
+	// series of nonces encoded in compressed format. It uses a slicing
+	// function to extra 33 bytes at a time from the packed 2x public
+	// nonces.
+	type nonceSlicer func([PubNonceSize]byte) []byte
+	combineNonces := func(slicer nonceSlicer) (*btcec.PublicKey, error) {
+		// Convert the set of nonces into jacobian coordinates we can
+		// use to accumulate them all into each other.
+		pubNonceJs := make([]*btcec.JacobianPoint, len(pubNonces))
+		for i, pubNonceBytes := range pubNonces {
+			// Using the slicer, extract just the bytes we need to
+			// decode.
+			var nonceJ btcec.JacobianPoint
+			pubNonce, err := btcec.ParsePubKey(
+				slicer(pubNonceBytes),
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			pubNonce.AsJacobian(&nonceJ)
+			pubNonceJs[i] = &nonceJ
+		}
+
+		// Now that we have the set of complete nonces, we'll aggregate
+		// them: R = R_i + R_i+1 + ... + R_i+n.
+		var aggregateNonce btcec.JacobianPoint
+		for _, pubNonceJ := range pubNonceJs {
+			btcec.AddNonConst(
+				&aggregateNonce, pubNonceJ, &aggregateNonce,
+			)
+		}
+
+		// Now that we've aggregated all the points, we need to check
+		// if this point is the point at infinity, if so, then we'll
+		// just return the generator. At a later step, the malicious
+		// party will be detected.
+		var infinityPoint btcec.JacobianPoint
```

Comment:
nit: this is later re-declared on the package level. Perhaps do that in this commit already and re-use it? Since it will be shadowed later.